src/SlidingWindow.cpp

```diff
@@ -78,10 +84,73 @@ class ExpandExpr : public IRMutator {
 Expr expand_expr(const Expr &e, const Scope<Expr> &scope) {
     ExpandExpr ee(scope);
     Expr result = ee.mutate(e);
-    debug(3) << "Expanded " << e << " into " << result << "\n";
+    debug(4) << "Expanded " << e << " into " << result << "\n";
     return result;
 }
 
+class FindProduce : public IRVisitor {
+    const string &func;
+
+    using IRVisitor::visit;
+
+    void visit(const ProducerConsumer *op) override {
+        if (op->is_producer && op->name == func) {
+            found = true;
+        } else {
+            IRVisitor::visit(op);
+        }
+    }
+
+public:
+    bool found = false;
+
+    FindProduce(const string &func)
+        : func(func) {
+    }
+};
+
+bool find_produce(const Stmt &s, const string &func) {
+    FindProduce finder(func);
+    s.accept(&finder);
+    return finder.found;
+}
+
+// Insert bounds on a dimension of a producer with a new min or max, or both.
+class GuardProducer : public IRMutator {
+    const Function &func;
+    int dim_idx;
+    // These may be undefined, indicating there is no bound.
+    const Expr &min;
+    const Expr &max;
+
+    using IRMutator::visit;
+
+    Stmt visit(const Provide *op) override {
+        if (op->name != func.name()) {
+            return op;
+        }
+        internal_assert(dim_idx < (int)op->args.size());
+        Expr var = op->args[dim_idx];
+        Expr guard = const_true();
+        if (min.defined()) {
+            guard = guard && likely_if_innermost(var >= min);
```

Comment:
The next simplifier pass is going to rewrite this to min <= var, so may as well write it that way.